### ISSUE TITLE
Deadlock in destructor, somewhere in channels

### ISSUE THREAD

**Issue by stepancheg** [tokio-rs/tokio#2058]
How to reproduce:

```
$ git clone https://github.com/stepancheg/tokio-repro-deadlock/
$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/server`
waiting for shutdown message
sending shutdown
joining the thread
thread done
# hangs forever here in destructor
```

Single file: https://github.com/stepancheg/tokio-repro-deadlock/blob/master/src/bin/server.rs

Reproducing example is quite complicated, could not make it smaller.

Does not reproduce when `features = ["full"]`.

Environment:
* macOS
```
% cargo --version; rustc --version
cargo 1.42.0-nightly (86134e766 2019-12-23)
rustc 1.42.0-nightly (1ed41b072 2020-01-01)
```

Versions:

* tokio=0.2.6
* futures=0.3.1

LLDB stack trace: https://gist.github.com/stepancheg/026cb0539f868293807ba4872c6a1d71

**Comment by appaquet** [comment]
Looks like the same issue I have reported https://github.com/tokio-rs/tokio/issues/2046

**Comment by carllerche** [comment]
Thanks for the detailed report. I believe that the issue is identified and a fix is in progress.

**Comment by carllerche** [comment]
I am no longer able to reproduce this on master. This is related to the single-threaded runtime deadlock fixed in 0.2.8.
